src/plasmapy/simulation/particle_tracker/particle_tracker.py

```diff
@@ -620,117 +703,205 @@ def particles_on_grid(self):
+        summed_field_values = {
+            field_name: np.zeros(self.nparticles_tracked)
+            * AbstractGrid.recognized_quantities[field_name].unit
+            for field_name in self._required_quantities
+        }
+
         for grid in self.grids:
-            # Estimate the E and B fields for each particle
-            # Note that this interpolation step is BY FAR the slowest part of the push
-            # loop. Any speed improvements will have to come from here.
             if self.field_weighting == "volume averaged":
-                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.volume_averaged_interpolator(
-                    pos_tracked * u.m,
-                    "E_x",
-                    "E_y",
-                    "E_z",
-                    "B_x",
-                    "B_y",
-                    "B_z",
-                    persistent=True,
+                # Use the keys of the storage dictionary as input quantity strings to the interpolator
+                field_values = grid.volume_averaged_interpolator(
+                    pos_tracked * u.m, *summed_field_values.keys(), persistent=True
                 )
             elif self.field_weighting == "nearest neighbor":
-                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.nearest_neighbor_interpolator(
-                    pos_tracked * u.m,
-                    "E_x",
-                    "E_y",
-                    "E_z",
-                    "B_x",
-                    "B_y",
-                    "B_z",
-                    persistent=True,
+                field_values = grid.nearest_neighbor_interpolator(
+                    pos_tracked * u.m, *summed_field_values.keys(), persistent=True
                 )
 
-            # Interpret any NaN values (points off the grid) as zero
-            # Do this before adding to the totals, because 0 + nan = nan
-            _Ex = np.nan_to_num(_Ex, nan=0.0 * u.V / u.m)
-            _Ey = np.nan_to_num(_Ey, nan=0.0 * u.V / u.m)
-            _Ez = np.nan_to_num(_Ez, nan=0.0 * u.V / u.m)
-            _Bx = np.nan_to_num(_Bx, nan=0.0 * u.T)
-            _By = np.nan_to_num(_By, nan=0.0 * u.T)
-            _Bz = np.nan_to_num(_Bz, nan=0.0 * u.T)
-
-            # Add the values interpolated for this grid to the totals
-            Ex += _Ex
-            Ey += _Ey
-            Ez += _Ez
-            Bx += _Bx
-            By += _By
-            Bz += _Bz
+            # Iterate through the interpolated fields and add them to the running sum
+            # NaN values are zeroed
+            for field_value, field_name in zip(
+                field_values, summed_field_values.keys(), strict=False
+            ):
+                summed_field_values[field_name] += np.nan_to_num(
+                    field_value,
+                    0.0 * AbstractGrid.recognized_quantities[field_name].unit,
+                )
 
         # Create arrays of E and B as required by push algorithm
         E = np.array(
```

Comment:
Maybe this should be done in the _update_position method, since it is just as input to the integrator? Then you can just return the summed fields variable from this fcn